### PR TITLE
Refactor old gens to for...of

### PR DESCRIPTION
**Pull request by KrisOnline**

### CODE REVIEW

**Comment by KrisOnline** [comment]
oHnO

File: mods/gen1/moves.js [hunk 1]
```diff
@@ -426,11 +425,9 @@ exports.BattleMovedex = {
 					if (pokemon.status === 'tox') {
 						pokemon.setStatus('psn');
 					}
-					let volatiles = Object.keys(pokemon.volatiles);
-					for (let n = 0; n < volatiles.length; n++) {
-						let id = volatiles[n];
+					for (const id of Object.keys(pokemon.volatiles)) {
 						if (id === 'residualdmg') {
-							pokemon.volatiles[id].counter = 0;
+							id.counter = 0;
```
Comment (Zarel) [inline mods/gen1/moves.js]: ...it doesn't work this way

Comment (KrisOnline) [reply mods/gen1/moves.js]: shhhhhh

Comment (Zarel) [reply mods/gen1/moves.js]: I'm actually surprised eslint does- oh I keep forgetting eslint doesn't know what types are.

Comment (Zarel) [reply mods/gen1/moves.js]: If you TypeScripted this file, though, TypeScript would be yelling at you here.

Comment (Zarel) [reply mods/gen1/moves.js]: Today is a good day to learn about what happens with primitive promotion in JavaScript! In JavaScript, there are a bunch of primitive types, but the main relevant ones here are `boolean`, `string`, and `number`. `id` is a `string`, specifically the string `"residualdmg"`.

So here's the thing: Primitives don't actually have attributes. Only objects have attributes.

When you access attributes on a primitive, JavaScript will temporarily promote the primitive to an object. So when you write something like:

```js
id.counter = 0;
```

What's actually going on behind the scenes is:

```js
{
  let promotedId = new Object(id);
  promotedId.counter = 0;
} // promotedId is immediately destroyed; your code had no effect
```

**Comment by KrisOnline** [comment]
oh whoops forgot to say something

i fixed the errors

File: mods/gen1/random-teams.js [hunk 2]
```diff
@@ -274,8 +273,8 @@ class RandomGen1Teams extends RandomGen2Teams {
 			if (movePool.length) {
 				hasMove = {};
 				counter = {Physical: 0, Special: 0, Status: 0, physicalsetup: 0, specialsetup: 0};
-				for (let k = 0; k < moves.length; k++) {
-					let move = this.getMove(moves[k]);
+				for (const i of moves) {
```
Comment (Zarel) [inline mods/gen1/random-teams.js]: I'd use `id` here

File: mods/gen1/rulesets.js [hunk 3]
```diff
@@ -15,8 +15,8 @@ exports.BattleFormats = {
 				problems.push(set.species + ' is not a real Pokemon.');
 			}
 			if (set.moves) {
-				for (let i = 0; i < set.moves.length; i++) {
-					let move = this.getMove(set.moves[i]);
+				for (const i of set.moves) {
```
Comment (Zarel) [inline mods/gen1/rulesets.js]: I'd use `id` here too

File: mods/gen1/rulesets.js [hunk 4]
```diff
@@ -62,12 +62,12 @@ exports.BattleFormats = {
...
 				let hasMove = {};
-				for (let i = 0; i < set.moves.length; i++) {
-					let move = this.getMove(set.moves[i]);
+				for (const i of set.moves) {
+					let move = this.getMove(i);
 					let moveid = move.id;
 					if (hasMove[moveid]) continue;
 					hasMove[moveid] = true;
-					moves.push(set.moves[i]);
+					moves.push(move);
```
Comment (Zarel) [inline mods/gen1/rulesets.js]: This is supposed to be `id`, not `move`

File: mods/gen1/rulesets.js [hunk 5]
```diff
@@ -62,12 +62,12 @@ exports.BattleFormats = {
 			let moves = [];
 			if (set.moves) {
 				let hasMove = {};
-				for (let i = 0; i < set.moves.length; i++) {
```
Comment (Zarel) [inline mods/gen1/rulesets.js]: `id` here.

File: mods/gen2/moves.js [hunk 6]
```diff
@@ -513,9 +513,9 @@ exports.BattleMovedex = {
 	roar: {
 		inherit: true,
 		onTryHit: function () {
-			for (let i = 0; i < this.queue.length; i++) {
+			for (const action of this.queue) {
 				// Roar only works if it is the last action in a turn, including when it's called by Sleep Talk
-				if (this.queue[i].choice === 'move' || this.queue[i].choice === 'switch') return false;
+				if (['move', 'switch'].includes(action.choice)) return false;
```
Comment (Zarel) [inline mods/gen2/moves.js]: Heyy, `includes` is a refactor for the old `in {object}` code. `||` is still more performant, and should be done for short tests like this one.

File: mods/gen2/moves.js [hunk 7]
```diff
@@ -719,9 +719,9 @@ exports.BattleMovedex = {
 	whirlwind: {
 		inherit: true,
 		onTryHit: function () {
-			for (let i = 0; i < this.queue.length; i++) {
+			for (const action of this.queue.length) {
 				// Whirlwind only works if it is the last action in a turn, including when it's called by Sleep Talk
-				if (this.queue[i].choice === 'move' || this.queue[i].choice === 'switch') return false;
+				if (['move', 'switch'].inludes(action.choice)) return false;
```
Comment (Zarel) [inline mods/gen2/moves.js]: Ditto `||`

File: mods/gen2/random-teams.js [hunk 8]
```diff
@@ -91,13 +91,12 @@ class RandomGen2Teams extends RandomGen3Teams {
...
-			for (let i = 0; i < moves.length; i++) {
-				if (restrictMoves[moves[i]]) restrictMoves[moves[i]]--;
-				if (restrictMoves['phazing'] && (moves[i] === "roar" || moves[i] === "whirlwind")) {
+			for (const move of set.moveset.moves) {
+				if (restrictMoves[move]) restrictMoves[move]--;
+				if (restrictMoves['phazing'] && ['roar', 'whirlwind'].includes(move)) {
 					restrictMoves['phazing']--;
 				}
-				if (restrictMoves['sleeping'] && (moves[i] === "sleeppowder" || moves[i] === "lovelykiss" || moves[i] === "sing" || moves[i] === "hypnosis" || moves[i] === "spore")) {
+				if (restrictMoves['sleeping'] && ['hypnosis', 'lovelykiss', 'sing', 'sleeppowder', 'spore'].includes(move)) {
```
Comment (Zarel) [inline mods/gen2/random-teams.js]: (This `includes` refactor is fine, though; four is definitely the point at which I'd switch to `includes`)

File: mods/gen2/random-teams.js [hunk 9]
```diff
@@ -91,13 +91,12 @@ class RandomGen2Teams extends RandomGen3Teams {
 
 			// Moves counter.
 			restrictMoves = set.other.restrictMoves;
-			let moves = set.moveset.moves;
-			for (let i = 0; i < moves.length; i++) {
-				if (restrictMoves[moves[i]]) restrictMoves[moves[i]]--;
-				if (restrictMoves['phazing'] && (moves[i] === "roar" || moves[i] === "whirlwind")) {
+			for (const move of set.moveset.moves) {
```
Comment (Zarel) [inline mods/gen2/random-teams.js]: Would use `moveid` here. I'm trying to be consistent by using `moveid` for move IDs and `move` for move objects.

File: mods/gen2/random-teams.js [hunk 10]
```diff
@@ -110,20 +109,20 @@ class RandomGen2Teams extends RandomGen3Teams {
 			}
 
 			// Type counter.
-			for (let i = 0; i < types.length; i++) {
-				if (typeCount[types[i]]) {
-					typeCount[types[i]]++;
+			for (const type of types) {
```
Comment (Zarel) [inline mods/gen2/random-teams.js]: Ditto `typeName`

Comment (Zarel) [reply mods/gen2/random-teams.js]: Actually, for types we can leave it alone.

File: mods/gen2/random-teams.js [hunk 11]
```diff
@@ -200,11 +199,11 @@ class RandomGen2Teams extends RandomGen3Teams {
 		let discourage = false;
 		if (!discard && slot > 3) {
 			discourage = true;
-			for (let i = 0; i < moves.length; i++) {
-				if (moves[i] === "sleeptalk" && restrictMoves['sleeptalk'] > 1) { discourage = false; break; }
-				if (moves[i] !== "bellydrum" && moves[i] !== "haze" && moves[i] !== "thief" && restrictMoves[moves[i]] > 0) { discourage = false; break; }
-				if (isPhazingMove(moves[i]) && restrictMoves['phazing'] > 0) { discourage = false; break; }
-				if (isSleepMove(moves[i]) && restrictMoves['sleeping'] > 1) { discourage = false; break; }
+			for (const move of moves) {
```
Comment (Zarel) [inline mods/gen2/random-teams.js]: `moveid`

File: mods/gen2/random-teams.js [hunk 12]
```diff
@@ -214,9 +213,9 @@ class RandomGen2Teams extends RandomGen3Teams {
 		if (set.item) item = set.item[this.random(set.item.length)];
 
 		// Adjust ivs for hiddenpower
-		for (let i = 0; i < moves.length; i++) {
-			if (moves[i].substr(0, 11) !== 'hiddenpower') continue;
-			let hpType = moves[i].substr(11, moves[i].length);
+		for (const move of moves) {
```
Comment (Zarel) [inline mods/gen2/random-teams.js]: `moveid`

File: mods/gen2/rulesets.js [hunk 13]
```diff
@@ -72,12 +72,12 @@ exports.BattleFormats = {
 			let moves = [];
 			if (set.moves) {
 				let hasMove = {};
-				for (let i = 0; i < set.moves.length; i++) {
-					let move = this.getMove(set.moves[i]);
+				for (const i of set.moves) {
```
Comment (Zarel) [inline mods/gen2/rulesets.js]: `id`

File: mods/gen2/rulesets.js [hunk 14]
```diff
@@ -24,8 +24,8 @@ exports.BattleFormats = {
 				}
 			}
 			if (set.moves) {
-				for (let i = 0; i < set.moves.length; i++) {
-					let move = this.getMove(set.moves[i]);
+				for (const i of set.moves) {
```
Comment (Zarel) [inline mods/gen2/rulesets.js]: `id`

File: mods/gen1/random-teams.js (hunk 2 above)
Comment (Zarel) [reply mods/gen1/random-teams.js]: In hindsight, all the places I've been saying `id`, we might want to use `setMoveid`

The main difference between `moveid` and `setMoveid` would be that `setMoveid` includes Hidden Power type information.

File: mods/gen5/random-teams.js [hunk 15]
```diff
@@ -51,8 +51,8 @@ class RandomGen5Teams extends RandomGen6Teams {
 			hasAbility[template.abilities['H']] = true;
 		}
 		let availableHP = 0;
-		for (let i = 0, len = movePool.length; i < len; i++) {
-			if (movePool[i].substr(0, 11) === 'hiddenpower') availableHP++;
+		for (const move of movePool) {
+			if (move.substr(0, 11) === 'hiddenpower') availableHP++;
```
Comment (Zarel) [inline mods/gen5/random-teams.js]: `move.startsWith(`

File: mods/gen5/random-teams.js [hunk 16]
```diff
@@ -51,8 +51,8 @@ class RandomGen5Teams extends RandomGen6Teams {
 			hasAbility[template.abilities['H']] = true;
 		}
 		let availableHP = 0;
-		for (let i = 0, len = movePool.length; i < len; i++) {
-			if (movePool[i].substr(0, 11) === 'hiddenpower') availableHP++;
+		for (const move of movePool) {
```
Comment (Zarel) [inline mods/gen5/random-teams.js]: `setMoveid`

File: mods/gen5/random-teams.js [hunk 17]
```diff
@@ -62,11 +62,11 @@ class RandomGen5Teams extends RandomGen6Teams {
 
 		do {
 			hasMove = {};
-			for (let k = 0; k < moves.length; k++) {
-				if (moves[k].substr(0, 11) === 'hiddenpower') {
+			for (const move of moves) {
+				if (move.substr(0, 11) === 'hiddenpower') {
```
Comment (Zarel) [inline mods/gen5/random-teams.js]: `setMoveid.startsWith(`

File: mods/gen5/random-teams.js [hunk 18]
```diff
@@ -84,8 +84,8 @@ class RandomGen5Teams extends RandomGen6Teams {
 
 			counter = this.queryMoves(moves, hasType, hasAbility, movePool);
 
-			for (let k = 0; k < moves.length; k++) {
-				let move = this.getMove(moves[k]);
+			for (const i of moves) {
```
Comment (Zarel) [inline mods/gen5/random-teams.js]: `for (const [i, setMoveid] of moves.entries()) {`

File: mods/gen5/random-teams.js [hunk 19]
```diff
@@ -313,7 +313,7 @@ class RandomGen5Teams extends RandomGen6Teams {
 
 				// Remove rejected moves from the move list
 				if (rejected && (movePool.length - availableHP || availableHP && (moveid === 'hiddenpower' || !hasMove['hiddenpower']))) {
-					moves.splice(k, 1);
+					moves.splice(moves.indexOf(move), 1);
```
Comment (Zarel) [inline mods/gen5/random-teams.js]: `i`

File: mods/gen5/random-teams.js [hunk 20]
```diff
@@ -514,8 +514,8 @@ class RandomGen5Teams extends RandomGen6Teams {
 		} else if (ability === 'Unburden' && (counter['Physical'] || counter['Special'])) {
 			// Give Unburden mons a random Gem of the type of one of their damaging moves
 			let eligibleTypes = [];
-			for (let i = 0; i < moves.length; i++) {
-				let move = this.getMove(moves[i]);
+			for (const i of moves) {
```
Comment (Zarel) [inline mods/gen5/random-teams.js]: `setMoveid`

Comment (Zarel) [reply mods/gen5/random-teams.js]: etc, just do it every time this comes up.

File: mods/gen6/random-teams.js [hunk 21]
```diff
@@ -61,8 +61,8 @@ class RandomGen6Teams extends RandomTeams {
 			hasAbility[template.abilities['H']] = true;
 		}
 		let availableHP = 0;
-		for (let i = 0, len = movePool.length; i < len; i++) {
-			if (movePool[i].substr(0, 11) === 'hiddenpower') availableHP++;
+		for (const move of movePool) {
+			if (move.substr(0, 11) === 'hiddenpower') availableHP++;
```
Comment (Zarel) [inline mods/gen6/random-teams.js]: Also `setMoveid.startsWith('hiddenpower')`, just do it every time this comes up, too.

File: mods/gen6/random-teams.js [hunk 22]
```diff
@@ -99,8 +99,8 @@ class RandomGen6Teams extends RandomTeams {
 			counter = this.queryMoves(moves, hasType, hasAbility, movePool);
 
 			// Iterate through the moves again, this time to cull them:
-			for (let k = 0; k < moves.length; k++) {
-				let move = this.getMove(moves[k]);
+			for (const i of moves) {
```
Comment (Zarel) [inline mods/gen6/random-teams.js]: `for (const [i, setMoveid] of moves.entries()) {`

File: mods/gen6/random-teams.js [hunk 23]
```diff
@@ -540,7 +540,7 @@ class RandomGen6Teams extends RandomTeams {
 
 				// Remove rejected moves from the move list
 				if (rejected && (movePool.length - availableHP || availableHP && (moveid === 'hiddenpower' || !hasMove['hiddenpower']))) {
-					moves.splice(k, 1);
+					moves.splice(moves.indexOf(move), 1);
```
Comment (Zarel) [inline mods/gen6/random-teams.js]: Literally never _add_ an `indexOf`, even if it means you have to stick to an old-style loop.

Comment (Zarel) [reply mods/gen6/random-teams.js]: (`i`)

**Comment by Zarel** [comment]
Everything else looks good.

**Comment by KrisOnline** [comment]
wtf

**Comment by KrisOnline** [comment]
i get it thank you

**Comment by KrisOnline** [comment]
woohoo i did it

**Comment by Zarel** [comment]
Cool, hopefully this doesn't come with any bugs...

**Pull request merged**